### PR TITLE
dns: removing exceptions from resolver

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low
Testing: updated tests
Docs Changes: n/a
Release Notes: n/a
https://github.com/envoyproxy/envoy-mobile/issues/176

### CODE REVIEW

**Comment by jmarantz** [comment]
Yanjun -- can you do a first pass? After your lgtm I'll assign a maintainer.

File: source/extensions/network/dns_resolver/cares/dns_impl.h [hunk 1]
```diff
@@ -45,10 +45,12 @@ class DnsResolverImplPeer;
...
 class DnsResolverImpl : public DnsResolver, protected Logger::Loggable<Logger::Id::dns> {
 public:
+  static absl::StatusOr<absl::optional<std::string>>
+  maybeBuildResolversCsv(const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers);
+
   DnsResolverImpl(
       const envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig& config,
-      Event::Dispatcher& dispatcher,
-      const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
+      Event::Dispatcher& dispatcher, absl::optional<std::string> resolvers_csv,
```
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.h:53]: Just curious what's the motivation for this change?

Comment (alyssawilk) [reply source/extensions/network/dns_resolver/cares/dns_impl.h:53]: envoyproxy/envoy-mobile#176 :-)

but more relevant to Envoy, there's also https://github.com/envoyproxy/envoy/issues/27412

File: source/extensions/network/dns_resolver/cares/dns_impl.cc [hunk 2]
```diff
@@ -557,26 +556,29 @@ class CaresDnsResolverFactory : public DnsResolverFactory,
...
       resolvers.reserve(resolver_addrs.size());
       for (const auto& resolver_addr : resolver_addrs) {
         auto address_or_error = Network::Address::resolveProtoAddress(resolver_addr);
-        THROW_IF_STATUS_NOT_OK(address_or_error, throw);
+        RETURN_IF_STATUS_NOT_OK(address_or_error);
         resolvers.push_back(std::move(address_or_error.value()));
       }
     }
-    return std::make_shared<Network::DnsResolverImpl>(cares, dispatcher, resolvers,
+    auto csv_or_error = DnsResolverImpl::maybeBuildResolversCsv(resolvers);
```
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:579]: It looks to me there are some logic change here.   Previously, if resolvers is empty, the code still go ahead create DnsResolverImpl:  https://github.com/envoyproxy/envoy/blob/0d1ce088b8d87e03bb0a3dcbeececdd97ad716f3/source/extensions/network/dns_resolver/cares/dns_impl.cc#L58. 

Now we skipped it.

Comment (alyssawilk) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:579]: I don't think so
we only fail if we have an error status.  in the case of empty  list, we still return absl::nullopt ("no resolvers") not absl::InvalidArgumentError  ("failure creating configured resolvers")

Comment (yanjunxiang-google) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:579]: I see. That static function DnsResolverImpl::maybeBuildResolversCsv() never returns Error.

But, it throw exceptions here:
https://github.com/envoyproxy/envoy/blob/9f494c0ba71f1a6d11864d8c4fc9afe298e4339a/source/extensions/network/dns_resolver/cares/dns_impl.cc#L67

Should we remove this throw?

Comment (alyssawilk) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:579]: yeah I had set up to remove it and not yet removed it.  Happy to do so (fingers crossed on unit tests, else I'll pick back up monday)

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 3]
```diff
@@ -581,21 +581,23 @@ RedisClusterFactory::createClusterWithConfig(
   if (!cluster.has_cluster_type() || cluster.cluster_type().name() != "envoy.clusters.redis") {
     return absl::InvalidArgumentError("Redis cluster can only created with redis cluster type.");
   }
+  auto resolver =
+      THROW_OR_RETURN_VALUE(selectDnsResolver(cluster, context), Network::DnsResolverSharedPtr);
```
Comment (yanjunxiang-google) [inline source/extensions/clusters/redis/redis_cluster.cc:585]: nit:  can we do same thing as other clusters here, like:

auto resolver_or_error = selectDnsResolver(cluster, context);
  THROW_IF_NOT_OK(resolver_or_error.status());

Then using std::move(resolver_or_error.value() in line 592.

**Comment by yanjunxiang-google** [comment]
LGTM modulo CI error.

**Review by yanjunxiang-google — APPROVED** [review]
LGTM

**Comment by jmarantz** [comment]
@ravenblackx for non-google review (just realized I don't need a senior maintainer since Alyssa is).

**Review by ravenblackx — APPROVED** [review]

File: source/extensions/clusters/redis/redis_cluster.cc (hunk 3 above)
Comment (ravenblackx) [reply source/extensions/clusters/redis/redis_cluster.cc:585]: I think `THROW_IF_NOT_OK` is significantly more readable than THROW_OR_RETURN_VALUE, too (i.e. consistency is nice, and my preference is for consistency to be achieved in that direction.)

**Comment by alyssawilk** [comment]
yeah I prefer THROW_OR_SET_VALUE to
foo_or_error = function
THROW_IF_STATUS_NOT_OK(foo_or_error.status()
foo = foo_or_error.value()
so I tend to use it when there's setters involved.

**Pull request merged**